_studio/mfx_lib/encode_hw/h264/src/mfx_h264_encode_hw.cpp

```diff
@@ -2393,30 +2516,53 @@ mfxStatus ImplementationAvc::AsyncRoutine(mfxBitstream * bs)
     if (m_stagesToGo & AsyncRoutineEmulator::STG_BIT_START_SCD)
     {
         MFX_AUTO_LTRACE(MFX_TRACE_LEVEL_HOTSPOTS, "Avc::STG_BIT_START_SCD");
+        DdiTask & task = m_ScDetectionStarted.back();
         if (IsExtBrcSceneChangeSupported(m_video))
         {
-            DdiTask & task = m_ScDetectionStarted.back();
             MFX_CHECK_STS(SCD_Put_Frame(task));
         }
         OnScdQueried();
     }
 
     if (m_stagesToGo & AsyncRoutineEmulator::STG_BIT_WAIT_SCD)
     {
+        DdiTask & task = m_ScDetectionFinished.front();
+
         MFX_AUTO_LTRACE(MFX_TRACE_LEVEL_HOTSPOTS, "Avc::STG_BIT_WAIT_SCD");
         if (IsExtBrcSceneChangeSupported(m_video))
         {
-            DdiTask & task = m_ScDetectionFinished.front();
+
             if (task.m_type[0] == 0)
                 task.m_type = GetFrameType(m_video, task.m_frameOrder - m_frameOrderIdrInDisplayOrder);
 
             MFX_CHECK_STS(Prd_LTR_Operation(task));
             MFX_CHECK_STS(SCD_Get_FrameType(task));
             AssignFrameTypes(task);
         }
-        OnScdFinished();
+        SubmitMCTF();
     }
 
+    if (m_stagesToGo & AsyncRoutineEmulator::STG_BIT_START_MCTF)
+    {
+        MFX_AUTO_LTRACE(MFX_TRACE_LEVEL_HOTSPOTS, "Avc::STG_BIT_START_MCTF");
+        if (IsMctfSupported(m_video))
+        {
+            DdiTask & task = m_MctfStarted.back();
+            MFX_CHECK_STS(SubmitToMctf(&task, amtScd.Get_frame_shot_Decision(), amtScd.Get_intra_frame_denoise_recommendation()));
```

Comment:
please do not put functions calls inside macro. It could lead to double call in macro expansion. You can use ```MFX_SAFE_CALL``` instead

Reply:
Done